stdlib/string.gr

```diff
@@ -519,3 +519,596 @@ export let endsWith = (p: String, s: String) => {
+          WasmI32.store8(bytes + bytesIdx + 1n, codePoint & 0xffn, _BYTES_OFFSET)
+          bytesIdx += 2n
+        } else {
+          // https://en.wikipedia.org/wiki/UTF-16#Code_points_from_U+010000_to_U+10FFFF
+          let uPrime = codePoint - 0x10000n
+          let w1 = ((uPrime & 0b11111111110000000000n) >>> 10n) + 0xD800n // High surrogate
+          let w2 = (uPrime & 0b00000000001111111111n) + 0xDC00n // Low surrogate
+          WasmI32.store8(bytes + bytesIdx, (w1 & 0xff00n) >>> 8n, _BYTES_OFFSET)
+          WasmI32.store8(bytes + bytesIdx + 1n, w1 & 0xffn, _BYTES_OFFSET)
+          WasmI32.store8(bytes + bytesIdx + 2n, (w2 & 0xff00n) >>> 8n, _BYTES_OFFSET)
+          WasmI32.store8(bytes + bytesIdx + 3n, w2 & 0xffn, _BYTES_OFFSET)
+          bytesIdx += 4n
+        }
+      },
+      UTF16_LE => {
+        let codePoint = getCodePoint(ptr)
+        if (codePoint <= 0xFFFFn) {
+          // <lo><hi>
+          WasmI32.store8(bytes + bytesIdx, codePoint & 0xffn, _BYTES_OFFSET)
+          WasmI32.store8(bytes + bytesIdx + 1n, (codePoint & 0xff00n) >>> 8n, _BYTES_OFFSET)
+          bytesIdx += 2n
+        } else {
+          // https://en.wikipedia.org/wiki/UTF-16#Code_points_from_U+010000_to_U+10FFFF
+          let uPrime = codePoint - 0x10000n
+          let w1 = ((uPrime & 0b11111111110000000000n) >>> 10n) + 0xD800n // High surrogate
+          let w2 = (uPrime & 0b00000000001111111111n) + 0xDC00n // Low surrogate
+          WasmI32.store8(bytes + bytesIdx, w1 & 0xffn, _BYTES_OFFSET)
+          WasmI32.store8(bytes + bytesIdx + 1n, (w1 & 0xff00n) >>> 8n, _BYTES_OFFSET)
+          WasmI32.store8(bytes + bytesIdx + 2n, w2 & 0xffn, _BYTES_OFFSET)
+          WasmI32.store8(bytes + bytesIdx + 3n, (w2 & 0xff00n) >>> 8n, _BYTES_OFFSET)
+          bytesIdx += 4n
+        }
+      },
+      UTF32_BE => {
+        let codePoint = getCodePoint(ptr)
+        WasmI32.store8(bytes + bytesIdx, (codePoint & 0xff000000n) >>> 24n, _BYTES_OFFSET)
+        WasmI32.store8(bytes + bytesIdx + 1n, (codePoint & 0xff0000n) >>> 16n, _BYTES_OFFSET)
+        WasmI32.store8(bytes + bytesIdx + 2n, (codePoint & 0xff00n) >>> 8n, _BYTES_OFFSET)
+        WasmI32.store8(bytes + bytesIdx + 3n, codePoint & 0xffn, _BYTES_OFFSET)
+        bytesIdx += 4n
+      },
+      UTF32_LE => {
+        let codePoint = getCodePoint(ptr)
+        WasmI32.store8(bytes + bytesIdx, codePoint & 0xffn, _BYTES_OFFSET)
+        WasmI32.store8(bytes + bytesIdx + 1n, (codePoint & 0xff00n) >>> 8n, _BYTES_OFFSET)
+        WasmI32.store8(bytes + bytesIdx + 2n, (codePoint & 0xff0000n) >>> 16n, _BYTES_OFFSET)
+        WasmI32.store8(bytes + bytesIdx + 3n, (codePoint & 0xff000000n) >>> 24n, _BYTES_OFFSET)
+        bytesIdx += 4n
+      },
+    };
+    ptr += n
+  }
+  WasmI32.toGrain(bytes): Bytes
+}
+
+// Byte->String decoding and helper functions:
+import Char from "char"
+
+@disableGC
+let writeUtf8CodePoint = (ptr, codePoint) => {
```

Comment:
This one also more or less exists in Char and could maybe have a shared utility.

Reply:
Per discussion offline, deferring to #684.